### PR TITLE
[move-prover] Towards transformation based architecture

### PR DESCRIPTION
**Pull request by wrwg**
This is a larger PR which brings us closer to a transformation based approach in the move prover, with the objective to move more of the logic in the Boogie backend into bytecode transformations.

The PR consists of a sequence of refactorings in the representation of expressions, bytecodes, and the transformation processor pipeline. Lots of this is boilerplate. It then adds a new processor [`spec_instrumentation.rs`](https://github.com/wrwg/diem/blob/by_trafo3/language/move-prover/bytecode/src/spec_instrumentation.rs) which translates a function specification into assert/assume bytecodes. This is not yet connected to anything and it is incomplete, but it demonstrates the direction in which this goes. The baseline test [here](https://github.com/wrwg/diem/tree/by_trafo3/language/move-prover/bytecode/tests/spec_instrumentation) is a good way to see the result.

The refactoring is split into multiple commits present in this PR:

1. Clean up expressions, specifically remove the hack of "virtual" operators to represent certain conditions like `aborts_with` and `modifies`.
2. Extend the transformation pipeline to allow to create multiple variants of a function. This is not used yet but will be needed later to have different code for the situation a function is verified and inlined. 
3. Finally implementation of the SpecInstrumenter and a new helper class FunctionDataBuilder, as well as full pretty printer for expressionn, as needed for the bytecode baseline testing infrastructure.

## Motivation

Towards transformation based architecture.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Existing tests + new ones.

## Related PRs

NA

### CODE REVIEW

**Comment by sblackshear** [comment]
Couple of questions:

- If a function has multiple control-flow paths that end in a `return` or `abort`, will the spec instrumentation add the postcondition/aborts_if asserts on every path?
- Will global specs also (eventually) be handled via a code transformation that adds asserts and assumes?

**Comment by wrwg** [comment]
> Couple of questions:
> 
> * If a function has multiple control-flow paths that end in a `return` or `abort`, will the spec instrumentation add the postcondition/aborts_if asserts on every path?

Right now yes (and abort isn't actually handled yet), but there is a transformation planned which will have a function have only one return instruction, to avoid duplication.

> * Will global specs also (eventually) be handled via a code transformation that adds asserts and assumes?

Yes, this is the plan. This is currently incomplete and probably only less than 50% of what needs to be done.

**Comment by JoelMarcey** [comment]
@oriooctopus Do you know why [this test](https://github.com/diem/diem/pull/7097/checks?check_run_id=1645416443) might be failing? It looks like it is having trouble finding some custom Dfuzr modules.... but I don't recall landing any Docusaurus changes over the holidays that would cause this error. I know you have a [PR](https://github.com/diem/diem/pull/7087) to upgrade Docusaurus to v2 0.7, but that hasn't landed yet.

```
./node_modules/libra-docusaurus-components/src/Layout/index.js
Module not found: Can't resolve '@theme/LayoutHead' in '/home/runner/work/diem/diem/developers.diem.com/node_modules/libra-docusaurus-components/src/Layout'./node_modules/libra-docusaurus-components/src/Layout/index.js
./node_modules/libra-docusaurus-components/src/Layout/index.js
Module not found: Can't resolve '@theme/LayoutProviders' in '/home/runner/work/diem/diem/developers.diem.com/node_modules/libra-docusaurus-components/src/Layout'./node_modules/libra-docusaurus-components/src/Layout/index.js
Module not found: Error: Can't resolve '@theme/LayoutHead' in '/home/runner/work/diem/diem/developers.diem.com/node_modules/libra-docusaurus-components/src/Layout'
 @ ./node_modules/libra-docusaurus-components/src/Layout/index.js 1:208-250 2:693-703
Module not found: Can't resolve '@theme/hooks/useKeyboardNavigation' in '/home/runner/work/diem/diem/developers.diem.com/node_modules/libra-docusaurus-components/src/Layout'
 @ ./node_modules/libra-docusaurus-components/src/index.js
Client bundle compiled with errors therefore further build is impossible.
 @ ./src/theme/DocItem/index.js
 @ ./.docusaurus/registry.js
 @ ./node_modules/@docusaurus/core/lib/client/exports/ComponentCreator.js
 @ ./.docusaurus/routes.js
 @ ./node_modules/@docusaurus/core/lib/client/clientEntry.js
 @ multi ./node_modules/@docusaurus/core/lib/client/clientEntry.js
./node_modules/libra-docusaurus-components/src/Layout/index.js
Module not found: Error: Can't resolve '@theme/LayoutProviders' in '/home/runner/work/diem/diem/developers.diem.com/node_modules/libra-docusaurus-components/src/Layout'
 @ ./node_modules/libra-docusaurus-components/src/Layout/index.js 1:250-302 2:580-595
 @ ./node_modules/libra-docusaurus-components/src/index.js
 @ ./src/theme/DocItem/index.js
 @ ./.docusaurus/registry.js
 @ ./node_modules/@docusaurus/core/lib/client/exports/ComponentCreator.js
 @ ./.docusaurus/routes.js
 @ ./node_modules/@docusaurus/core/lib/client/clientEntry.js
 @ multi ./node_modules/@docusaurus/core/lib/client/clientEntry.js
./node_modules/libra-docusaurus-components/src/Layout/index.js
Module not found: Error: Can't resolve '@theme/hooks/useKeyboardNavigation' in '/home/runner/work/diem/diem/developers.diem.com/node_modules/libra-docusaurus-components/src/Layout'
 @ ./node_modules/libra-docusaurus-components/src/Layout/index.js 1:349-419 2:517-538
 @ ./node_modules/libra-docusaurus-components/src/index.js
 @ ./src/theme/DocItem/index.js
 @ ./.docusaurus/registry.js
 @ ./node_modules/@docusaurus/core/lib/client/exports/ComponentCreator.js
 @ ./.docusaurus/routes.js
 @ ./node_modules/@docusaurus/core/lib/client/clientEntry.js
 @ multi ./node_modules/@docusaurus/core/lib/client/clientEntry.js
 ```

**Review by shazqadeer — COMMENTED** [review]
In the spec instrumentation output for Increment_R, I noticed that @0 was used to save memory but @0 was not declared as a local.  Is there some special status of these variables into which things are saved?

**Comment by wrwg** [comment]
> In the spec instrumentation output for Increment_R, I noticed that @0 was used to save memory but @0 was not declared as a local. Is there some special status of these variables into which things are saved?

This is not a local because it might be the entire memory R not just at some point (tau, addr). So we can't represent it like this. I'm not sure yet that we need to declare those memory snapshots in some way, for now I assume no.

**Review by shazqadeer — APPROVED** [review]

File: language/move-prover/bytecode/src/stackless_bytecode.rs [hunk 1]
```diff
@@ -239,6 +191,10 @@ pub enum Bytecode {
     Label(AttrId, Label),
     Abort(AttrId, TempIndex),
     Nop(AttrId),
+
+    SaveMem(AttrId, MemoryLabel, QualifiedId<StructId>),
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:195]: Yes

**Pull request closed without merging**